Implement Popen Output Streamer

rmartin16 opened the pull request:
As detailed in #676, the attempt to display the iOS simulator log can miss the first few seconds of log output. This is because the capturing of the log output is started after the simulator is started; therefore, there will always be a non-zero amount of time that log output is not being captured.

Log streaming for macOS and iOS simulator now start before the app is started; once the app is running, any queued log output is flushed to the screen and new log output is printed in real time.

## Implementation
 - A new function in `subprocess.py` loops on the `stdout` output of an established `Popen` process
   - The loop will properly exit when it detects the underlying process has exited
   - The loop will properly exit and terminate the underlying process if the user sends CTRL+C
 - Starting up the iOS simulator was updated to initiate `log stream` via `Popen`
 - Once the simulator is started, the new output streaming function is called to print all of the simulator log

## PR Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
- [x] All new features have been tested
- [x] All new features have been documented
- [x] I have read the **CONTRIBUTING.md** file
- [x] I will abide by the code of conduct

rmartin16 commented:
@freakboy3742, before I move too much further forward with this proposal, can you please provide any thoughts you have?

I know you mentioned threading in #676 but I felt that `Popen` would give us better control over this (especially since we're already doing so much with `subprocess`). This approach makes it more straightforward to start showing output exactly when we want; and in the most general use case for this, I thought this was superior over kicking off a command in to another thread where it would definitely be trickier to control when the thread started printing....especially if we wanted to print more things in the main thread first.

freakboy3742 commented on src/briefcase/platforms/iOS/xcode.py:
```diff
@@ -399,6 +402,22 @@ def run_app(self, app: BaseConfig, udid=None, **kwargs):
+                "xcrun", "simctl", "spawn", udid,
+                "log", "stream",
+                "--style", "compact",
+                "--predicate", 'senderImagePath ENDSWITH "/{app.formal_name}"'.format(app=app)
```
This can now be an f-string

rmartin16 replied:
Converted to an f-string.

freakboy3742 commented on src/briefcase/platforms/iOS/xcode.py:
```diff
@@ -421,19 +440,9 @@ def run_app(self, app: BaseConfig, udid=None, **kwargs):
-            )
+            stream_popen_output(simulator_log_popen, self.logger.info)
+        except OutputStreamingError as e:
+            raise BriefcaseCommandError(f"Unable to start log stream for app {app.app_name}: {e}")
```
To my reading, there's a couple of conditions other than "unable to start" that could trigger this case; this means we either need a more errors, or a more generic error.

rmartin16 replied:
That's true....although, I changed this up and expounded in the next post.

freakboy3742 reviewed (changes requested):
Oooh... I like this a *lot*. Popen streaming is a *much* better approach than threading - in hindsight, it's so obvious that I'm kicking myself for not thinking of it.

Other than the things I've flagged inline the only thing that stands out is that `stream_popen_output` will likely need to be another augmentation on the subprocess module so that it can be easily intercepted for testing purposes. (i.e., `subprocess.stream_output(process)`).

rmartin16 commented on src/briefcase/platforms/iOS/xcode.py:
```diff
@@ -338,6 +341,21 @@ def run_app(self, app: BaseConfig, udid=None, **kwargs):
+            stderr=subprocess.STDOUT,
+            bufsize=1,
+        )
+        time.sleep(3)  # let log stream start up
```
I'm not sure how long to wait here....ideally, i suppose, it'd be nice to know when `log stream` is ready....but I don't think that's an option...

rmartin16 replied:
So, on my exceptionally slow virtualized instance of macOS, I did encounter an instance of missing the beginning of the log when this wait was 1 second.

While starting `log stream` before running the app at least makes it possible to display the entire log, I believe there is still a race condition. When `Popen` is invoked, it simply forks a process with instructions to run a command; so, I'm not sure there would necessarily be any guarantee of the state of that command once `Popen` returns control.

So, if `log stream` (for some bizarre reason) takes 10 seconds to start actually capturing the log entries, it's quite likely that the app will already be up and running. That's why, for instance, it would be nice if we could enter another loop sending `SIGUSR1` signals or something to `log stream` until it responded affirmatively that it's ready to actually stream logs. Then we'd know 100% it is safe to start the app and show all the log entries.

side note: i found this in `Popen` ;) thought it somewhat apt
```python
        # How long to resume waiting on a child after the first ^C.
        # There is no right value for this.  The purpose is to be polite
        # yet remain good for interactive users trying to exit a tool.
        self._sigint_wait_secs = 0.25  # 1/xkcd221.getRandomNumber()
```

freakboy3742 replied:
Ah - that makes sense. I guess we need to make the call between slow enough to guarantee that *really* slow machines will have started the log, and being slow enough to get in the way of app development. 

My gut tells me that 3 seconds is leaning a little to hard on the "accomodate slow machines" side; made worse by the fact that it's a 3 second wait before any other console output happens (so it's *right* on the cusp of a "did that work?" wait). If you're running macOS virtualised, you're definitely going to be seeing the worst case scenario, and most people doing macOS development won't be doing so in a virtualised container. I wonder if the 0.25 value you've found actually might be good enough in this case...

rmartin16 replied:
I found that `0.25` seconds did work sometimes on my virtualized macOS...maybe 50/50. As you said though, this should definitely be the extreme example of delay on macOS....I can only even allocate a single CPU to the VM without a kernel panic :| so I'm fine with `0.25`....can always increase later.

rmartin16 commented:
I moved the stream function inside of `Subprocess` still taking a `Popen` process. However, I removed any sort of detection of "failure" from it. The more I thought about it, the more it felt out of place to try to interpret if the process was in a state of failure just because it had seemingly exited prior to the stream function being invoked. Instead, the stream function will look for output, print it, and exit as normal.

That said, if we wanted to detect failure, it seems more appropriate to do so back up where streaming is being invoked. However, even here, I was having trouble determining when `log stream` may have failed....or if it ran normally but looks like it failed. For instance, if I close the iOS simulator while `log stream` is still running, the `log stream` process is killed automatically and returns a non-zero code. However, this isn't really a failure mode but looks very similar to what might be considered one. (Replicating this on `main` results in a "unable to start log stream" error.) Finally, I'm not even sure if it's worth tracking if `log stream` itself failed....it seems other problems would likely show up elsewhere before `log stream` failed. That said, we can add checks on its return codes perhaps if you think it's worth it.

I implemented this log streaming for the iOS simulator and macOS apps. AFAICT, adb's `logcat` will print everything in its ring buffer log that matches the filter....so starting it early doesn't seem to offer a benefit. Linux and Windows apps seem to depend on normal printing `stdout` for logging. That said, let me know if I should add this anywhere else you know of.

rmartin16 commented on src/briefcase/integrations/subprocess.py:
```diff
@@ -206,6 +210,39 @@ def Popen(self, args, **kwargs):
+            except subprocess.TimeoutExpired:
+                popen_process.kill()
+
+        except Exception as e:
```
On one hand, this sort of catching feels somewhat overdone seems Im really not sure how `readline` or `poll` can error....on the other hand, re-wrapping whatever might come out in a more contextual error seems appropriate.

rmartin16 commented on src/briefcase/integrations/subprocess.py:
```diff
@@ -206,6 +210,39 @@ def Popen(self, args, **kwargs):
+            except subprocess.TimeoutExpired:
+                popen_process.kill()
+
+        except Exception as e:
+            raise PopenStreamingError(f"{e!r}")
```
On one hand, this sort of catching feels somewhat overdone seems Im really not sure how readline or poll can error....on the other hand, re-wrapping whatever might come out in a more contextual error seems appropriate.

rmartin16 replied:
That's a good perspective. Removed the catch.

rmartin16 commented on src/briefcase/platforms/macOS/__init__.py:
```diff
@@ -32,6 +34,35 @@ def run_app(self, app: BaseConfig, **kwargs):
+            stderr=subprocess.STDOUT,
+            bufsize=1,
+        )
+        time.sleep(3)  # let log stream start up
```
same here

rmartin16 replied:
Left a comment about timeout above. I'll comment on `Popen` exceptions here.

I think that catching `Popen` exceptions is probably unnecessary. As I detailed the behavior of `run` in another comment, `run` is a wrapper on `Popen`. In that analysis, it became clear to me that if `Popen` itself raised an exception (sans `TimeoutError`), `run` would let that exception bubble up in to the caller's code. In all the calling code here, we only ever catch `CalledProcessError` for `run` which will only occur for non-zero return codes (and if `check=True`).

So, given this and the fact `Popen` exceptions are discovered at dev time (via invalid arguments) or reveal a deep OS level runtime issue, I think it makes sense not to try to catch any exceptions from `Popen`.

freakboy3742 replied:
I agree it's unlikely to fail; but if it does, we have an opportunity to fail gracefully - in a way that lets the user know it's probably not their fault, and in a way that maximises the likelihood that their bug report will point at the actual cause of the problem. 

Consider - in this case, in the "caught" scenario, the bug report is "Briefcase is raising an error saying it can't start the log stream". In the "uncaught" scenario, the bug report is a `subprocess.CalledProcessError`, and if we're *really* lucky, they'll provide the full stack trace from which we might be able to work back and determine it's the log stream start that has failed. 

Catching the error and reporting the high-level problem is defensive programming on our part.

freakboy3742 replied:
Huh... for some reason I thought Popen generated other errors (akin to `run()`) , but checking the docs, you're right - it's a fairly limited set of errors that would firmly fall in the "what the hell have you done to your computer?!?" edge case. On that basis, what you've got here is probably sufficient; if the user has borked their system enough that they're getting permission errors out of `log`, surfacing the raw error is probably OK.

freakboy3742 reviewed (changes requested):
This is looking good. I've flagged a few things; mostly around ensuring error handling and flow paths are obvious. Some of them might be stemming from my own misunderstanding; if that's the case, it's probably an indication that more explanatory comments are needed.

I agree with your analysis that trying to detect failure is more trouble than it's worth; as indicated inline, we should verify that the process *starts*, but once it has started, I don't know that there's much we can meaningfully detect. 

I think you're right that this is only needed on iOS and macOS at present; Android adb logging certainly seems to be able to functionally capture "all" the logs. Linux logging is a direct stdout/stderr stream, so there's nothing extra needed; and until we fix #629, there isn't any log to stream on Windows.

rmartin16 commented on src/briefcase/integrations/subprocess.py:
```diff
@@ -206,6 +210,39 @@ def Popen(self, args, **kwargs):
+                return_code = popen_process.poll()
+                if output_line == "" and return_code is not None:
+                    self._log_return_code(return_code)
+                    break
```
breaking an `if` huh....ngl, though, `break`s can definitely be problematic...especially in nested code. thanks for sharing.

Updated to `return`.

rmartin16 commented on src/briefcase/integrations/subprocess.py:
```diff
@@ -206,6 +210,39 @@ def Popen(self, args, **kwargs):
+            try:
+                popen_process.wait(timeout=3)
+            except subprocess.TimeoutExpired:
+                popen_process.kill()
```
Good call. Added a warning log.

rmartin16 commented on tests/integrations/subprocess/test_Subprocess__stream_output.py:
```diff
@@ -0,0 +1,68 @@
+
+
+def test_output_debug(mock_sub, popen_process, capsys):
+    "Readline output is printed"
```
There isn't a difference. I added this test only to confirm that "debug" mode is the same as "non-debug" mode. Should only see a difference in "deep debug". Clarified that in the description....although, I suppose these tests could also be combined.

rmartin16 commented on tests/integrations/subprocess/test_Subprocess__stream_output.py:
```diff
@@ -0,0 +1,68 @@
+
+def test_keyboard_interrupt(mock_sub, popen_process):
+    "Process is terminated if user sends CTRL+C"
+    popen_process.stdout.readline.side_effect = KeyboardInterrupt()
```
Good point. Updated.

rmartin16 commented on src/briefcase/integrations/subprocess.py:
```diff
@@ -206,6 +210,39 @@ def Popen(self, args, **kwargs):
+                # empty string, that should mean the underlying process is exiting/gone
+                output_line = ensure_str(popen_process.stdout.readline())
+                return_code = popen_process.poll()
+                if output_line == "" and return_code is not None:
```
I was a little leery about this loop as well at first. However, I believe it hinges on assumptions that are already baked in to our other uses of `subprocess`.

- `check_output` is wrapper on `run` and `run` is a wrapper on `Popen`.
- `Popen` does some basic validation, sets up `stdin/out/err`, and then forks a process to run the OS command.
  - Strictly speaking, `Popen` can fail but it appears to be either invalid input combinations or something went really wrong at the OS level.
- From here, `run` blocks on `read\write` calls to `stdin/out/err` via `Popen.communicate()`.
  - These `read` call implementations are buried in layers of abstraction, but all the documentation of python `io` suggests that `read` returns when it hits an `EOF`.
  - For `readline`, the docs say that once an `EOF` is hit, `readline` will only ever return `""` or `b""` for subsequent calls.
  - For `stdout/err`, I am interpreting `EOF` to mean that the child process closed the pipes on its end to them.
- Once `Popen.communicate()` returns, `run` blocks on `wait()`.
  - True to its name, it waits until the child process finishes and provides a return code.
- If return code is non-zero, raise `CalledProcessError`
- Alternatively, return a `CompletedProcess` object with return code and `stdout/err`

So....in this way, this loop is a really dumbed down version of `run`, in that:
- At some point, `readline()` will need to start returning `""` because the process closed the pipe
- And the child process will need to return an code via `poll()`

If these assumptions aren't met from a child process, then we'll need the user to send CTRL+C.....however, that would be true even using `run` (at least without a `timeout` since we don't provide one).


As for the specific structure of the loop:
 - The quit condition is based on both `readline()` returning empty and `poll()` providing a return code.
 - This is mostly to prevent a race condition with flushing `readline` output to the screen and `poll()` getting a return code from the process.
 - Requiring both means we've flushed all the output and we have a return code.
 - That said, I think there are other configurations for the loop, but I don't think they fundamentally change the need for `readline` returns `""` and `poll()` is not `None`.
 - Finally, this `Popen` loop idiom was surprisingly popular when I was researching the best way to do this....although, it's always hard to tell if that's because everyone just copy/pastes the seemingly most popular thing...or if its actually the best way to do this.

freakboy3742 replied:
Ok - this all makes sense, and the revised loop structure looks good to me. `poll()` is definitely less of a risk than `wait()` under the circumstances; and AFAICT the usage of `poll()` you've got here should be safe.

rmartin16 commented:
Responded to comments.

Outstanding items:
 - Structure of `Popen` output streaming loop
 - Waiting after starting `log stream`
 - `Popen` exception catching

freakboy3742 reviewed (changes requested):
Getting very close; comments inline - at least one of which is a "🤷‍♂️... how much will be Enough?".

rmartin16 commented:
I reduced the wait time for `log stream` to `0.25` seconds.

I also responded inline about the `Popen` exceptions....just want to double check.

freakboy3742 reviewed (approved):
Looks great! I picked up one last detail while doing a close check of the tests - in the event of the app failing to start, the log process would be left dangling. I'm fairly certain this would be cleaned up when the parent process dies; but it doesn't hurt to do the cleanup, and it was easy to factor out - plus it gave an opportunity to provide slightly more contextual cleanup messages.

Thanks for another great contribution!
